src/Fixer/Semicolon/MultilineWhitespaceBeforeSemicolonsFixer.php

```diff
@@ -0,0 +1,292 @@
+    {
+        return new FixerConfigurationResolver([
+            (new FixerOptionBuilder(
+                    'strategy',
+                    'Forbid multi-line whitespace or move the semicolon to the new line for chained calls.'
+            ))
+                ->setAllowedValues([self::STRATEGY_NO_MULTI_LINE, self::STRATEGY_NEW_LINE_FOR_CHAINED_CALLS])
+                ->setDefault(self::STRATEGY_NO_MULTI_LINE)
+                ->getOption(),
+        ]);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    protected function applyFix(\SplFileInfo $file, Tokens $tokens)
+    {
+        if (self::STRATEGY_NEW_LINE_FOR_CHAINED_CALLS === $this->configuration['strategy']) {
+            return $this->applyChainedCallsFix($tokens);
+        }
+
+        if (self::STRATEGY_NO_MULTI_LINE === $this->configuration['strategy']) {
+            return $this->applyNoMultiLineFix($tokens);
+        }
+    }
+
+    /**
+     * @param Tokens $tokens
+     */
+    protected function applyNoMultiLineFix(Tokens $tokens)
+    {
+        $lineEnding = $this->whitespacesConfig->getLineEnding();
+
+        foreach ($tokens as $index => $token) {
+            if (!$token->equals(';')) {
+                continue;
+            }
+
+            $previousIndex = $index - 1;
+            $previous = $tokens[$previousIndex];
+            if (!$previous->isWhitespace() || false === strpos($previous->getContent(), "\n")) {
+                continue;
+            }
+
+            $content = $previous->getContent();
+            if (("\n" === $content[0] || "\r" === $content[0]) && $tokens[$index - 2]->isComment()) {
+                $tokens[$previousIndex] = new Token([$previous->getId(), $lineEnding]);
+            } else {
+                $tokens->clearAt($previousIndex);
+            }
+        }
+    }
+
+    /**
+     * @param Tokens $tokens
+     */
+    protected function applyChainedCallsFix(Tokens $tokens)
+    {
+        for ($index = count($tokens) - 1; $index >= 0; --$index) {
+            // continue if token is not a semicolon
+            if (!$tokens[$index]->equals(';')) {
+                continue;
+            }
+
+            // get the indent of the chained call, null in case it's not a chained call
+            $indent = $this->getIndentation($index - 1, $tokens);
+
+            if (null === $indent) {
+                continue;
+            }
+
+            // unset semicolon
+            $tokens->clearAt($index);
+
+            // find the line ending token index after the semicolon
+            $index = $this->getNewLineIndex($index, $tokens);
+
+            // line ending string of the last method call
+            $lineEnding = $this->whitespacesConfig->getLineEnding();
+
+            // appended new line to the last method call
+            $newline = new Token([T_WHITESPACE, $lineEnding.$indent]);
+
+            // insert the new line with indented semicolon
+            $tokens->insertAt($index, [$newline, new Token(';')]);
```

Comment:
I'm not sure, but it looks like there is unset token, then set back.

What happens if you keep semicolon and just insert whitespace before it?

```php
// $tokens->clearAt($index);

// ... the same

$tokens->insertAt($index, [$newline]);
```

Reply:
True, the semicolon is unset. Inserting just the white space would break this use case https://github.com/FriendsOfPHP/PHP-CS-Fixer/pull/3327/files#diff-705ca1a8f094c9edea7f841893060cdaR196

Reply:
I see, thanks